dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/impl/NettyQuiesceListener.java

```diff
@@ -0,0 +1,138 @@
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-2.0/
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package io.openliberty.netty.internal.impl;
+
+import java.util.Set;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import com.ibm.websphere.ras.Tr;
+import com.ibm.websphere.ras.TraceComponent;
+
+import io.netty.channel.Channel;
+import io.netty.channel.group.ChannelGroup;
+
+
+public class NettyQuiesceListener{
+	
+	/** Trace service*/
+	private static final TraceComponent tc = Tr.register(NettyQuiesceListener.class, NettyConstants.NETTY_TRACE_NAME, NettyConstants.BASE_BUNDLE);
+	
+	private ScheduledExecutorService scheduler;
+	
+	private ScheduledFuture<?> quiesceFinishTask;
+	private ScheduledFuture<?> quiesceVerifierTask;
+	private long quiesceTimeout = 0;
+	
+	private Runnable quiesceTask = null;
+
+	private NettyFrameworkImpl bundle;
+	
+	/**
+	 * Constructor
+	 */
+	public NettyQuiesceListener(NettyFrameworkImpl nettyBundle, ScheduledExecutorService scheduler, long timeout) {
+		this.bundle = nettyBundle;
+		this.scheduler = scheduler;
+		this.quiesceTimeout = (timeout > 0) ? timeout:0;
+	}
+			
+
+	/**
+	 * Runs the Quiesce verifier task at fixed intervals until the active channels have
+	 * been stopped or the Quiesce finish task finishes. The Quiesce finish task will 
+	 * be ran once after the Quiesce time has elapsed. 
+	 * @return
+	 */
+	public ScheduledFuture<?> startTasks() {
+		
+		this.quiesceFinishTask = scheduler.schedule( ()->{
+			if(TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+            	Tr.debug(tc, "The quiesceFinishTask has started for " + this.bundle + ". Quiesce task will be scheduled to run for: " + this.bundle);
+            }
+			if(quiesceTask != null)
+				scheduler.schedule(quiesceTask, 0, TimeUnit.MILLISECONDS);
+			if(!this.quiesceVerifierTask.isDone()) {
+				if(TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+	            	Tr.debug(tc, "The quiesceVerifierTask was found not to be done. Cancelling it...");
+	            }
+				this.quiesceVerifierTask.cancel(true);
+			}
+		}, quiesceTimeout, TimeUnit.MILLISECONDS);
+		
+		// TODO: Maybe we should check the map for the active channels if they have any connections left that could be better
+		
+		this.quiesceVerifierTask = scheduler.scheduleAtFixedRate(() -> {
+			if(TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+            	Tr.debug(tc, "The quiesceVerifierTask has started,  checking endpoints...");
+            }
+			Set<Channel> activeChannels = this.bundle.getActiveChannels();
+			if(activeChannels.isEmpty()) {
```

Comment:
In this case, connections refer to the active connections to the endpoints (child channels started from the main server channel). Since channels refer to pretty much any connection on Netty, I used this terminology to distinguish them: `Connections` for channels created from a ServerChannel or outbound connections, and `Endpoints` for TCP channels which are bind to a port. How can I make this more understandable? Maybe adding some comments to make sure they don't get confused?

Reply:
I will happily adopt the same terminology if/as it makes sense, it can be good to inoculate such conventions with comments but I will catch up either way and try to be consistent.